### PR TITLE
[BEAM-10961] Enabled strict dependency on Spark

### PR DESCRIPTION
**Pull request by sonam-vend**
Since enabling the strict dependencies on all Java modules is a large and critical change to be merged all at once. Therefore, task [[BEAM-10961](https://issues.apache.org/jira/browse/BEAM-10961)] is divided into sub-tasks and strict dependencies are enabled to each module separately which is comparatively easy to review and merge.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by sonam-vend** [comment]
R: @ibzib , @aaltay can you please review or add reviewers who can review this.

**Comment by aaltay** [comment]
Sonam, the number of PRs are too many. Could you create a spreadsheet of these changes and we can try to find owners? Perhaps we can advertise the list on the dev list and see who can help with what?

@ibzib - If you are familiar with any parts please start reviewing those.
@kennknowles @tysonjh - There seems to be 10-20 PRs like this. Do you have a suggestion on best way to handle reviews?

**Comment by ibzib** [comment]
> Sonam, the number of PRs are too many. Could you create a spreadsheet of these changes and we can try to find owners? Perhaps we can advertise the list on the dev list and see who can help with what?
> 
> @ibzib - If you are familiar with any parts please start reviewing those.
> @kennknowles @tysonjh - There seems to be 10-20 PRs like this. Do you have a suggestion on best way to handle reviews?

@aaltay For context, it was my suggestion to split #12938 into many PRs, since that was such a large change that affected our entire build. I'm guessing some review comments will be applicable across many PRs, but I still think it's best to merge projects individually so we don't "break the world." I will start by reviewing the sections I am most familiar with.

File: runners/spark/build.gradle [hunk 1]
```diff
@@ -84,6 +84,12 @@ dependencies {
   compile library.java.slf4j_api
   compile library.java.joda_time
   compile library.java.args4j
+  compile project(path: ":model:fn-execution", configuration: "shadow")
+  compile project(path: ":model:job-management:", configuration: "shadow")
+  compile project(":sdks:java:fn-execution")
+  compile library.java.vendored_grpc_1_26_0
+  compile library.java.vendored_guava_26_0_jre
+  compile "org.checkerframework:checker-qual:2.0.0"
```
Comment (kennknowles) [inline runners/spark/build.gradle:92]: This dependency was suggested to be added in BeamModulePlugin. Or at least it should have a shared definition there. It is a universal Java dependency.

Comment (sonam-vend) [reply runners/spark/build.gradle:92]: @kennknowles ,  "org.checkerframework:checker-qual" has a universal definition in https://github.com/apache/beam/pull/13690/files#diff-0435a83a413ec063bf7e682cadcd56776cd18fc878f197cc99a65fc231ef2047

Comment (kennknowles) [reply runners/spark/build.gradle:92]: I see. So it should not be needed here. Perhaps we should very soon merge just those changers to the universal definitions and then update the smaller PRs.

File: runners/spark/job-server/build.gradle [hunk 2]
```diff
@@ -49,14 +49,12 @@ configurations.all {
...
 
 dependencies {
-  compile project(sparkRunnerProject)
   compile project(path: sparkRunnerProject, configuration: "provided")
   validatesPortableRunner project(path: sparkRunnerProject, configuration: "testRuntime")
   validatesPortableRunner project(path: sparkRunnerProject, configuration: "provided")
   validatesPortableRunner project(path: ":sdks:java:core", configuration: "shadowTest")
   validatesPortableRunner project(path: ":runners:core-java", configuration: "testRuntime")
   validatesPortableRunner project(path: ":runners:portability:java", configuration: "testRuntime")
-  compile project(":sdks:java:extensions:google-cloud-platform-core")
```
Comment (ibzib) [inline runners/spark/job-server/build.gradle:59]: I believe we need this as a runtime-only dependency for the expansion service.

**Review by ibzib — CHANGES_REQUESTED** [review]

File: runners/spark/job-server/build.gradle (hunk 2 above)
Comment (sonam-vend) [reply runners/spark/job-server/build.gradle:59]: corrected,thanks!

File: runners/spark/build.gradle (hunk 1 above)
Comment (sonam-vend) [reply runners/spark/build.gradle:92]: @kennknowles sure, https://github.com/apache/beam/pull/13690/ PR is ready for review.

File: runners/spark/build.gradle [hunk 3]
```diff
@@ -84,6 +84,12 @@ dependencies {
   compile library.java.slf4j_api
   compile library.java.joda_time
   compile library.java.args4j
+  compile project(path: ":model:fn-execution", configuration: "shadow")
+  compile project(path: ":model:job-management:", configuration: "shadow")
```
Comment (ibzib) [inline runners/spark/build.gradle:88]:
```suggestion
  compile project(path: ":model:job-management", configuration: "shadow")
```

File: runners/spark/build.gradle (hunk 1 above)
Comment (ibzib) [reply runners/spark/build.gradle:92]: #13690 is merged, can you update this now?

Comment (sonam-vend) [reply runners/spark/build.gradle]: > #13690 is merged, can you update this now?

removed from spark build.gradle as it is reference there in Universal deps.

**Comment by ibzib** [comment]
Run Java Spark PortableValidatesRunner Batch

**Comment by ibzib** [comment]
Run Spark ValidatesRunner

**Comment by ibzib** [comment]
Run Python Spark ValidatesRunner

**Comment by ibzib** [comment]
Run Java Spark PortableValidatesRunner Batch

**Review by ibzib — COMMENTED** [review]
@sonam-vend shouldn't we enable strict dependencies on each module as we go, instead of waiting for every module to be fixed to start checking any of them?

**Comment by shehzaadn-vd** [comment]
> @sonam-vend shouldn't we enable strict dependencies on each module as we go, instead of waiting for every module to be fixed to start checking any of them?
 
@ibzib Since #12938 is passing all checks, and all these PRs were split from that one, @sonam-vend and I were thinking we'd have a small-ish number of failures when we enable the switch at the global level. Since eventually we'll be turning on the global switch anyway, it'll help us avoid changes at the module level which we would then have to un do. Wdyt?

**Comment by sonam-vend** [comment]
Run Python Spark ValidatesRunner

**Comment by sonam-vend** [comment]
Run Python Spark ValidatesRunner

**Comment by sonam-vend** [comment]
Run Python Spark ValidatesRunner

**Comment by sonam-vend** [comment]
Run Python Spark ValidatesRunner

**Comment by sonam-vend** [comment]
Run Python Spark ValidatesRunner

**Comment by sonam-vend** [comment]
Run Python Spark ValidatesRunner

**Comment by sonam-vend** [comment]
Run Python Spark ValidatesRunner Batch

**Comment by sonam-vend** [comment]
Run Python Spark ValidatesRunner

**Comment by sonam-vend** [comment]
Run Python Spark ValidatesRunner

**Comment by ibzib** [comment]
> @sonam-vend and I were thinking we'd likely have only a small-ish number of issues (hopefully zero or thereabouts!) when we enable the switch at the global level.

@shehzaadn-vd Why hope for zero issues when we could guarantee zero issues?

> Since eventually we'll be turning on the global switch anyway, enabling strict dependencies at the module level would have to be undone anyway.

It's a trivial amount of work to add and remove these settings. The benefit of gaining strict dependency analysis on each module immediately outweighs the cost of copy+paste+deleting a few lines in each file.

File: runners/spark/job-server/build.gradle [hunk 4]
```diff
@@ -49,14 +49,13 @@ configurations.all {
 }
 
 dependencies {
-  compile project(sparkRunnerProject)
```
Comment (ibzib) [inline runners/spark/job-server/build.gradle:52]: @sonam-vend All Spark Python tests are failing. I suspect this line is the cause.

**Comment by shehzaadn-vd** [comment]
> > @sonam-vend and I were thinking we'd likely have only a small-ish number of issues (hopefully zero or thereabouts!) when we enable the switch at the global level.
> 
> @shehzaadn-vd Why hope for zero issues when we could guarantee zero issues?
> 
> > Since eventually we'll be turning on the global switch anyway, enabling strict dependencies at the module level would have to be undone anyway.
> 
> It's a trivial amount of work to add and remove these settings. The benefit of gaining strict dependency analysis on each module immediately outweighs the cost of copy+paste+deleting a few lines in each file.

@ibzib Understood. Let us look into that.

**Comment by ibzib** [comment]
> > > @sonam-vend and I were thinking we'd likely have only a small-ish number of issues (hopefully zero or thereabouts!) when we enable the switch at the global level.
> > 
> > 
> > @shehzaadn-vd Why hope for zero issues when we could guarantee zero issues?
> > > Since eventually we'll be turning on the global switch anyway, enabling strict dependencies at the module level would have to be undone anyway.
> > 
> > 
> > It's a trivial amount of work to add and remove these settings. The benefit of gaining strict dependency analysis on each module immediately outweighs the cost of copy+paste+deleting a few lines in each file.
> 
> @ibzib Understood. Let us look into that.

All you need to do is add `enableStrictDependencies: true` as an argument to `applyJavaNature` in each build.gradle file.

**Comment by sonam-vend** [comment]
> > > > @sonam-vend and I were thinking we'd likely have only a small-ish number of issues (hopefully zero or thereabouts!) when we enable the switch at the global level.
> > > 
> > > 
> > > @shehzaadn-vd Why hope for zero issues when we could guarantee zero issues?
> > > > Since eventually we'll be turning on the global switch anyway, enabling strict dependencies at the module level would have to be undone anyway.
> > > 
> > > 
> > > It's a trivial amount of work to add and remove these settings. The benefit of gaining strict dependency analysis on each module immediately outweighs the cost of copy+paste+deleting a few lines in each file.
> > 
> > 
> > @ibzib Understood. Let us look into that.
> 
> All you need to do is add `enableStrictDependencies: true` as an argument to `applyJavaNature` in each build.gradle file.

@ibzib , should I too push the section:
applyJavaNature(
  enableStrictDependencies: true,
//whatsoever
)

or is it just for the validity in my local for each module?

**Comment by ibzib** [comment]
> @ibzib , should I too push the section:
> applyJavaNature(
> enableStrictDependencies: true,
> //whatsoever
> )
> 
> or is it just for the validity in my local for each module?

Please push it so that future changes by other developers are checked.

File: runners/spark/job-server/build.gradle [hunk 5]
```diff
@@ -50,13 +50,14 @@ configurations.all {
```
Comment (ibzib) [inline runners/spark/job-server/build.gradle:50]: You'll need to set `enableStrictDependencies` in this file too.

**Comment by sonam-vend** [comment]
Run Java PreCommit

**Comment by ibzib** [comment]
Run Python Spark ValidatesRunner

**Review by ibzib — APPROVED** [review]
LGTM since this is PR is an improvement. However I believe there is something we will need to solve later: BEAM-11663. It's not a blocker for this PR, since this PR is not causing any regression, it is only revealing existing problems.

**Pull request merged**